util/cloudenv/cloudenv.go

```diff
@@ -66,20 +81,88 @@ func Get() Cloud {
 	return c
 }
 
+func readFileTrimmed(name string) string {
+	v, _ := os.ReadFile(name)
+	return strings.TrimSpace(string(v))
+}
+
 func getCloud() Cloud {
-	// TODO(bradfitz): also detect AWS on Windows, etc. Just try to hit the metadata server
-	// and see if it's there? But it might be turned off. Do some small-timeout DNS request
-	// to 169.254.169.253 and see if it replies? But which DNS request?
-	if runtime.GOOS == "linux" {
-		biosVendorB, _ := os.ReadFile("/sys/class/dmi/id/bios_vendor")
-		biosVendor := strings.TrimSpace(string(biosVendorB))
+	var hitMetadata bool
+	switch runtime.GOOS {
+	case "android", "ios", "darwin":
+		// Assume these aren't running on a cloud.
+		return ""
+	case "linux":
+		biosVendor := readFileTrimmed("/sys/class/dmi/id/bios_vendor")
 		if biosVendor == "Amazon EC2" || strings.HasSuffix(biosVendor, ".amazon") {
 			return AWS
 		}
+
+		prod := readFileTrimmed("/sys/class/dmi/id/product_name")
+		if prod == "Google Compute Engine" {
+			return GCP
+		}
+		if prod == "Google" { // old GCP VMs, it seems
+			hitMetadata = true
+		}
+		if prod == "Virtual Machine" || biosVendor == "Microsoft Corporation" {
+			// Azure, or maybe all Hyper-V?
+			hitMetadata = true
+		}
+	default:
+		// TODO(bradfitz): use Win32_SystemEnclosure from WMI or something on
+		// Windows to see if it's a physical machine and skip the cloud check
+		// early. Otherwise use similar clues as Linux about whether we should
+		// burn up to 2 seconds waiting for a metadata server that might not be
+		// there. And for BSDs, look where the /sys stuff is.
+		return ""
 	}
-	if gcpmetadata.OnGCE() {
+	if !hitMetadata {
+		return ""
+	}
+
+	const maxWait = 2 * time.Second
+	tr := &http.Transport{
+		DisableKeepAlives: true,
+		Dial: (&net.Dialer{
+			Timeout: maxWait,
+		}).Dial,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), maxWait)
+	defer cancel()
+
+	req, _ := http.NewRequestWithContext(ctx, "HEAD", "http://"+CommonNonRoutableMetadataIP, nil)
```

Comment:
I think this turns into a IMDSv1 request, a GET (HEAD, in this case) to the local Metadata server. IMDSv1 is disabled in many AWS deployments now and, more problematically, may have CloudWatch configured to watch for IMDSv1 requests.

Reply:
AWS IMDSv1 and AWS IMDSv2 and GCP metadata and Azure metadata are all four just HTTP to this same IP address.

To use IMDSv2 you need to make an HTTP request like this without any token to get a token. (`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 60"`)

The CloudWatch `MetadataNoToken` metric is mentioned at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/viewing_metrics_with_cloudwatch.html but doesn't say what exactly it counts. I confidently assume it doesn't count requests to get a token. I less confidently assume it shouldn't count no-op requests like a `HEAD /` that aren't trying to do anything, but who knows.

I could try to get a token instead, but that might just fire a CloudWatch-like counter on Azure or GCP instead.

I suppose I could experiment with CloudWatch and see if it triggers with this.

Looking at their SDK,

https://docs.aws.amazon.com/sdk-for-go/api/aws/ec2metadata/#EC2Metadata.Available
```
// Available returns if the application has access to the EC2 Metadata service.
// Can be used to determine if application is running within an EC2 Instance and
// the metadata service is available.
func (c *EC2Metadata) Available() bool {
```

.... it fetches a token (which would've been enough to answer the question of whether EC2Metadata was available) and then proceeds to fetch the instance-id using that token.

So we could fetch a token I guess?

But look a few lines up: on Linux, we don't even make it this far if we already found from the /sys info that we're on EC2. So this path would only trigger for Windows in the future (currently Windows returns false for all clouds)

Reply:
Well, ....

```
$ for x in $(seq 0 500); do curl  -I http://169.254.169.254/; done
....
```
<img width="630" alt="Screen Shot 2022-06-30 at 8 50 06 AM" src="https://user-images.githubusercontent.com/2621/176722218-eddade0a-ff6b-48ec-84d7-a82260593f29.png">

That answers that question.

Reply:
But no spike from:

```
ubuntu@bradprod:~$ for x in $(seq 0 500); do curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 5" ; done
```

<img width="785" alt="Screen Shot 2022-06-30 at 9 23 55 AM" src="https://user-images.githubusercontent.com/2621/176728817-c0761cf8-cf20-4c76-8d78-d84bada9b4c6.png">

Reply:
@DentonGentry, PTAL.